packages/client/src/actions/applyAction.ts

```diff
@@ -16,58 +16,93 @@
 
 import type { ActionDefinition } from "@osdk/api";
 import type { DataValue } from "@osdk/omniapi";
-import { applyActionV2 } from "@osdk/omniapi/OntologiesV2_Action";
+import {
+  applyActionBatchV2,
+  applyActionV2,
+} from "@osdk/omniapi/OntologiesV2_Action";
 import type { MinimalClient } from "../MinimalClientContext.js";
 import { addUserAgent } from "../util/addUserAgent.js";
 import { toDataValue } from "../util/toDataValue.js";
 import type {
   ActionEditResponse,
   ActionValidationResponse,
   ApplyActionOptions,
+  ApplyBatchActionOptions,
   OsdkActionParameters,
 } from "./Actions.js";
 import { ActionValidationError } from "./ActionValidationError.js";
 
 // cannot specify both validateOnly and returnEdits as true
 
-export type ActionReturnTypeForOptions<Op extends ApplyActionOptions> =
-  Op extends { validateOnly: true } ? ActionValidationResponse
-    : Op extends { returnEdits: true } ? ActionEditResponse
-    : undefined;
+export type ActionReturnTypeForOptions<
+  Op extends ApplyActionOptions | ApplyBatchActionOptions,
+> = Op extends { validateOnly: true } ? ActionValidationResponse
+  : Op extends { returnEdits: true } ? ActionEditResponse
+  : undefined;
 
 export async function applyAction<
   AD extends ActionDefinition<any, any>,
-  Op extends ApplyActionOptions,
+  P extends OsdkActionParameters<AD["parameters"]> | OsdkActionParameters<
+    AD["parameters"]
+  >[],
+  Op extends P extends OsdkActionParameters<
+    AD["parameters"]
+  >[] ? ApplyBatchActionOptions
+    : ApplyActionOptions,
 >(
   client: MinimalClient,
   action: AD,
-  parameters?: OsdkActionParameters<AD["parameters"]>,
+  parameters?: P,
   options: Op = {} as Op,
-): Promise<ActionReturnTypeForOptions<Op>> {
-  const response = await applyActionV2(
-    addUserAgent(client, action),
-    client.ontologyRid,
-    action.apiName,
-    {
-      parameters: remapActionParams(parameters),
-      options: {
-        mode: options?.validateOnly ? "VALIDATE_ONLY" : "VALIDATE_AND_EXECUTE",
-        returnEdits: options?.returnEdits ? "ALL" : "NONE",
+): Promise<
+  ActionReturnTypeForOptions<Op>
+> {
+  if (Array.isArray(parameters)) {
+    const response = await applyActionBatchV2(
+      addUserAgent(client, action),
+      client.ontologyRid,
+      action.apiName,
+      {
+        requests: parameters ? remapBatchActionParams(parameters) : [],
+        options: {
+          returnEdits: options?.returnEdits ? "ALL" : "NONE",
+        },
       },
-    },
-  );
+    );
 
-  if (options?.validateOnly) {
-    return response.validation as ActionReturnTypeForOptions<Op>;
-  }
+    return (options?.returnEdits
+      ? response.edits
+      : undefined) as ActionReturnTypeForOptions<Op>;
+  } else {
+    const response = await applyActionV2(
+      addUserAgent(client, action),
+      client.ontologyRid,
+      action.apiName,
+      {
+        parameters: remapActionParams(
+          parameters as OsdkActionParameters<AD["parameters"]>,
```

Comment:
So I had to cast here because even though I did the `Array.isArray` check above, it wasn't able to tell that this would be the non array type I defined above.